src/category_theory/limits/exact_functor.lean

```diff
@@ -0,0 +1,135 @@
+
+/-- Bundled exact functors. -/
+@[derive category, nolint has_inhabited_instance]
+def ExactFunctor :=
+{ F : C ⥤ D // nonempty (preserves_finite_limits F) ∧ nonempty (preserves_finite_colimits F) }
+
+infixr ` ⥤ₑ `:26 := ExactFunctor
+
+/-- An exact functor is in particular a functor. -/
+@[derive full, derive faithful]
+def ExactFunctor.forget : (C ⥤ₑ D) ⥤ (C ⥤ D) :=
+full_subcategory_inclusion _
+
+/-- Turn an exact functor into a left exact functor. -/
+@[derive full, derive faithful]
+def LeftExactFunctor.of_exact : (C ⥤ₑ D) ⥤ (C ⥤ₗ D) :=
+full_subcategory.map (λ X, and.left)
+
+/-- Turn an exact functor into a left exact functor. -/
+@[derive full, derive faithful]
+def RightExactFunctor.of_exact : (C ⥤ₑ D) ⥤ (C ⥤ᵣ D) :=
+full_subcategory.map (λ X, and.right)
+
+variables {C D}
+
+@[simp] lemma LeftExactFunctor.of_exact_obj (F : C ⥤ₑ D) :
+  (LeftExactFunctor.of_exact C D).obj F = ⟨F.1, F.2.1⟩ := rfl
+@[simp] lemma RightExactFunctor.of_exact_obj (F : C ⥤ₑ D) :
+  (RightExactFunctor.of_exact C D).obj F = ⟨F.1, F.2.2⟩ := rfl
+
+@[simp] lemma LeftExactFunctor.of_exact_map {F G : C ⥤ₑ D} (α : F ⟶ G) :
+  (LeftExactFunctor.of_exact C D).map α = α := rfl
+@[simp] lemma RightExactFunctor.of_exact_map {F G : C ⥤ₑ D} (α : F ⟶ G) :
+  (RightExactFunctor.of_exact C D).map α = α := rfl
+
+@[simp] lemma LeftExactFunctor.forget_obj (F : C ⥤ₗ D) :
+  (LeftExactFunctor.forget C D).obj F = F.1 := rfl
+@[simp] lemma RightExactFunctor.forget_obj (F : C ⥤ᵣ D) :
+  (RightExactFunctor.forget C D).obj F = F.1 := rfl
+@[simp] lemma ExactFunctor.forget_obj (F : C ⥤ₑ D) :
+  (ExactFunctor.forget C D).obj F = F.1 := rfl
+
+@[simp] lemma LeftExactFunctor.forget_map {F G : C ⥤ₗ D} (α : F ⟶ G) :
+  (LeftExactFunctor.forget C D).map α = α := rfl
+@[simp] lemma RightExactFunctor.forget_map {F G : C ⥤ᵣ D} (α : F ⟶ G) :
+  (RightExactFunctor.forget C D).map α = α := rfl
+@[simp] lemma ExactFunctor.forget_map {F G : C ⥤ₑ D} (α : F ⟶ G) :
+  (ExactFunctor.forget C D).map α = α := rfl
+
+/-- Turn a left exact functor into an object of the category `LeftExactFunctor C D`. -/
+def LeftExactFunctor.of (F : C ⥤ D) [preserves_finite_limits F] : C ⥤ₗ D := ⟨F, ⟨infer_instance⟩⟩
+/-- Turn a right exact functor into an object of the category `RightExactFunctor C D`. -/
+def RightExactFunctor.of (F : C ⥤ D) [preserves_finite_colimits F] : C ⥤ᵣ D :=
+⟨F, ⟨infer_instance⟩⟩
+/-- Turn an exact functor into an object of the category `ExactFunctor C D`. -/
+def ExactFunctor.of (F : C ⥤ D) [preserves_finite_limits F] [preserves_finite_colimits F] :
+  C ⥤ₑ D := ⟨F, ⟨⟨infer_instance⟩, ⟨infer_instance⟩⟩⟩
+
+@[simp] lemma LeftExactFunctor.of_fst (F : C ⥤ D) [preserves_finite_limits F] :
+  (LeftExactFunctor.of F).1 = F := rfl
+@[simp] lemma RightExactFunctor.of_fst (F : C ⥤ D) [preserves_finite_colimits F] :
+  (RightExactFunctor.of F).1 = F := rfl
+@[simp] lemma ExactFunctor.of_fst (F : C ⥤ D) [preserves_finite_limits F]
+  [preserves_finite_colimits F] : (ExactFunctor.of F).1 = F := rfl
```

Comment:
Would it be possible/desirable here (and in the instances below) to avoid using the `.1` projection, and instead put in the forgetful functor (or a coercion)?

I'm not sure it is necessary, just wondering if later you will be happier to have hidden the implementation of `LeftExactFunctor` as a subtype.

Reply:
The main reason I did it this way is that `Sheaf` does it this way. Maybe @erdOne can give more details about why this choice was made?

Reply:
I'm not sure about the problems exactly, but maybe [this Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Redefining.20Sheaves) is relevant?

Reply:
Okay, I just got very confused because sheaves on a site follow the structure approach while sheaves on a topological space follow the subtype approach. Thanks for the link, I wasn't aware of this discussion.

Reply:
Full subcategories now have moved to a custom structure and for now we use the probjection `.obj` to access the underlying object, as that seems to work well so far.